src/serializer/functions.js

```diff
@@ -103,25 +107,39 @@ export class Functions {
       if (property) {
         let value = property.descriptor && property.descriptor.value;
 
-        if (
-          !(
-            value instanceof FunctionValue ||
-            (value instanceof AbstractValue && valueIsKnownReactAbstraction(this.realm, value))
-          )
-        ) {
-          invariant(value instanceof AbstractValue);
-          realm.handleError(
-            new CompilerDiagnostic(
-              `Additional Function Value ${funcId} is an AbstractValue which is not allowed (unless a React known abstract)`,
-              undefined,
-              "PP0001",
-              "FatalError"
-            )
-          );
-          throw new FatalError("Additional Function values cannot be AbstractValues");
+        if (value instanceof ECMAScriptSourceFunctionValue) {
+          // additional function logic
+          recordedAdditionalFunctions.set(value, { funcId });
+          continue;
+        } else if (value instanceof ObjectValue) {
+          // React component tree logic
+          let config = Get(realm, value, "config");
+          let rootComponent = Get(realm, value, "rootComponent");
+          let validConfig = config instanceof ObjectValue || config === realm.intrinsics.undefined;
+          let validRootComponent =
+            rootComponent instanceof ECMAScriptSourceFunctionValue ||
+            (rootComponent instanceof AbstractValue && valueIsKnownReactAbstraction(this.realm, rootComponent));
+
+          if (validConfig && validRootComponent) {
+            recordedAdditionalFunctions.set(((rootComponent: any): ECMAScriptSourceFunctionValue | AbstractValue), {
+              funcId,
+              config: convertConfigObjectToReactComponentTreeConfig(
+                realm,
+                ((config: any): ObjectValue | UndefinedValue)
+              ),
+            });
+          }
+          continue;
         }
-        invariant(value instanceof AbstractValue || value instanceof ECMAScriptSourceFunctionValue);
-        recordedAdditionalFunctions.set(value, funcId);
+        realm.handleError(
+          new CompilerDiagnostic(
+            `Additional Function Value ${funcId} is an invalid value`,
+            undefined,
+            "PP0001",
```

Comment:
This seems like a new error message that we need to add to the wiki. Probably shouldn’t take the number one spot for this one.

Reply:
It's the same error though as PP001 that already exists.

Reply:
PP001 is used for errors that still need to be converted to proper diagnostics. You should never use it in new code.

Reply:
I copied and pasted the code from: https://github.com/facebook/prepack/blob/master/src/serializer/functions.js#L135-L139

So we should change that site too? @cblappert

Reply:
Additional function related messages should be in the PP1000 range.